v1/tarball/layer.go

```diff
@@ -64,33 +64,40 @@ func (l *layer) Size() (int64, error) {
 
 // LayerFromFile returns a v1.Layer given a tarball
 func LayerFromFile(path string) (v1.Layer, error) {
-	file, err := os.Open(path)
+	opener := func() (io.ReadCloser, error) {
+		return os.Open(path)
+	}
+	return LayerFromOpener(opener)
+}
+
+// LayerFromOpener returns a v1.Layer given an Opener function
+func LayerFromOpener(opener Opener) (v1.Layer, error) {
+	rc, err := opener()
 	if err != nil {
 		return nil, err
 	}
 
-	defer file.Close()
+	defer rc.Close()
 
-	compressed, err := v1util.IsGzipped(file)
+	compressed, err := v1util.IsGzipped(rc)
 	if err != nil {
 		return nil, err
 	}
 
-	if _, err := file.Seek(0, io.SeekStart); err != nil {
+	if rc, err = opener(); err != nil {
```

Comment:
This overwrites `rc` with a fresh copy, so either the first or this will be closed (IDK which the `defer` semantics will pick, but probably the first?).

I think the cleanest solution would be to simply pass the opener into `computeDigest` and `computeDiffID` instead of a `Reader`.

Reply:
I'm testing the reviewer edit functionality to fix this, since I want it too :)